src/charm.py

```diff
@@ -9,19 +9,132 @@
 """
 
 import logging
+import os
 from typing import Any
 
 import ops
+import yaml
+from ops import CollectStatusEvent, ConfigChangedEvent, RelationChangedEvent, UpgradeCharmEvent
+from ops.model import ActiveStatus, BlockedStatus, WaitingStatus
 
 logger = logging.getLogger(__name__)
 
+CLOUDS_YAML = "/etc/openstack/clouds.yaml"
+# used to match entry in clouds.yaml with openstack-exporter config
+CLOUD_NAME = "openstack"
+
 
 class OpenstackExporterOperatorCharm(ops.CharmBase):
     """Charm the service."""
 
     def __init__(self, *args: tuple[Any]) -> None:
         """Initialize the charm."""
         super().__init__(*args)
+        self.framework.observe(self.on.upgrade_charm, self._on_upgrade)
+        self.framework.observe(self.on.config_changed, self._on_config_changed)
+        self.framework.observe(self.on.credentials_relation_changed, self._on_credentials_changed)
+        self.framework.observe(self.on.collect_unit_status, self._on_collect_unit_status)
+
+    def _on_upgrade(self, _: UpgradeCharmEvent) -> None:
+        """Reconfigure everything on upgrade."""
+        self._configure()
```

Comment:
This is unnecessary, please remove it. Charm upgrade will run config event https://juju.is/docs/sdk/upgrade-charm-event